sharktank/sharktank/tools/tuner/candidate_gen.py

```diff
@@ -0,0 +1,1386 @@
+        self.tile_dims = tile_dims
+        self.mlir_template = mlir_template
+
+    def supports(self, mlir: str) -> bool:
+        return "matmul_like" in mlir
+
+    def get_candidate_gen_fn(self) -> CandidateGenFn:
+        if is_broadcast_rhs_mmt(self.mlir_template):
+            get_shapes_fn = get_shapes_broadcast_rhs_mmt
+            apply_params_fn = apply_params_broadcast_rhs_mmt
+        else:
+            get_shapes_fn = lambda template: get_shapes_contract(
+                template, self.lhs_dims, self.rhs_dims
+            )
+            apply_params_fn = lambda ps, template, config: apply_params_contract(
+                ps, self.tile_dims, template, config
+            )
+        return CandidateGenFn(get_shapes_fn, apply_params_fn)
+
+
+class BatchMmtTuner(DispatchTuner):
+    def supports(self, mlir: str) -> bool:
+        return "batch_matmul_transpose_b" in mlir
+
+    def get_candidate_gen_fn(self) -> CandidateGenFn:
+        return CandidateGenFn(get_shapes_batch_mmt, apply_params_batch_mmt)
+
+
+class BatchMatmulTuner(DispatchTuner):
+    def __init__(self, lhs_dims: str, rhs_dims: str):
+        self.lhs_dims = lhs_dims
+        self.rhs_dims = rhs_dims
+
+    def supports(self, mlir: str) -> bool:
+        return "batch_matmul" in mlir
+
+    def get_candidate_gen_fn(self) -> CandidateGenFn:
+        get_shapes_fn = lambda template: get_shapes_batch_matmul(
+            template, self.lhs_dims, self.rhs_dims
+        )
+        apply_params_fn = lambda ps, template, config: apply_params_batch_matmul(
+            ps, self.tile_dims, template, config
+        )
+        return CandidateGenFn(get_shapes_fn, apply_params_fn)
+
+
+def walk_callback_get_fn(
+    op: ir.Operation,
+    candidate_gen_fn: CandidateGenFn,
+    dispatch_tuner_registry: DispatchTunerRegistry,
+) -> ir.WalkResult:
+    if op.name == "util.func":
+        func_name = str(op.opview.sym_name)
+        searched_fn = dispatch_tuner_registry.get_candidate_gen_fn(func_name)
+        candidate_gen_fn.get_shapes_fn = searched_fn.get_shapes_fn
+        candidate_gen_fn.apply_params_fn = searched_fn.apply_params_fn
+        if candidate_gen_fn.apply_params_fn and candidate_gen_fn.get_shapes_fn:
+            return ir.WalkResult.INTERRUPT
+    return ir.WalkResult.ADVANCE
+
+
+def walk_mlir_op(
+    mlir_module: ir.Module,
+    candidate_gen_fn: CandidateGenFn,
+    dispatch_tuner_registry: DispatchTunerRegistry,
+):
+    for op in mlir_module.body.operations:
+        op.walk(
+            lambda op: walk_callback_get_fn(
+                op, candidate_gen_fn, dispatch_tuner_registry
+            ),
+            ir.WalkOrder.POST_ORDER,
+        )
+        if candidate_gen_fn.apply_params_fn and candidate_gen_fn.get_shapes_fn:
+            break
+
+
+def tune(
+    input: str,
+    output: str = "",
+    limit: int = 4096,
+    num_subgroups: int = 4,
+    lhs_dims: str = "mk",
+    rhs_dims: str = "nk",
+    tile_dims: str = "mnk",
+):
```

Comment:
As this is the main entrypoint function to this script, can you add some docstring comments for each argument?